### ISSUE TITLE
failed call to cuInit: CUDA_ERROR_UNKNOWN after Docker build on Macbook Pro (Late 2013) with Linux

### ISSUE THREAD

**Issue by dcunited001** [tensorflow/tensorflow#601]
This issue is similar to to #394. I believe i'm seeing it because tensorflow can't find `libcuda.so`.

I know that my `libcuda.so` can be found in `/usr/lib/x86_64-linux-gnu/`, but is this accessible to the compiler and to python without further configuration?  I tried adding it to my docker user's `$LD_LIBRARY_PATH`, but that did not fix the problem. 

I'm running on:
- Macbook Pro (Late 2013) Hardware
- Mint 17
- 3.0 Compute Capability
- CUDA 7.0 & cuDNN 2.0 
- running docker service under a docker user & group.  no sudo access for docker user.

on host os, values for `$CUDA_SO` and `$DEVICES` that are passed into `./docker_run_gpu.sh` are:

```
$ export CUDA_SO=$(\ls /usr/lib/x86_64-linux-gnu/libcuda* | xargs -I{} echo '-v {}:{}')
$ export DEVICES=$(\ls /dev/nvidia* | xargs -I{} echo '--device {}:{}')

$ echo $CUDA_SO
-v /usr/lib/x86_64-linux-gnu/libcuda.so:/usr/lib/x86_64-linux-gnu/libcuda.so -v /usr/lib/x86_64-linux-gnu/libcuda.so.1:/usr/lib/x86_64-linux-gnu/libcuda.so.1 -v /usr/lib/x86_64-linux-gnu/libcuda.so.352.68:/usr/lib/x86_64-linux-gnu/libcuda.so.352.68

$ echo $DEVICES
--device /dev/nvidia0:/dev/nvidia0 --device /dev/nvidiactl:/dev/nvidiactl
```

I customized `Dockerfile.devel-gpu` as below.  I prepended `TF_CUDA_COMPUTE_CAPABILITIES=3.0` and  `TF_UNOFFICIAL_SETTING=1` for the `./configure` step.  

Should I add the `libcuda.so` path to `ENV LD_LIBRARY_PATH`?

``` Dockerfile
# ........

RUN git clone --recursive https://github.com/tensorflow/tensorflow.git && \
    cd tensorflow && \
    git checkout 0.6.0
WORKDIR /tensorflow

# Configure the build for our CUDA configuration.
ENV CUDA_TOOLKIT_PATH /usr/local/cuda
ENV CUDNN_INSTALL_PATH /usr/local/cuda
ENV TF_NEED_CUDA 1

RUN TF_CUDA_COMPUTE_CAPABILITIES=3.0 TF_UNOFFICIAL_SETTING=1 ./configure && \
    bazel build -c opt --config=cuda tensorflow/tools/pip_package:build_pip_package && \
    bazel-bin/tensorflow/tools/pip_package/build_pip_package /tmp/pip && \
    pip install --upgrade /tmp/pip/tensorflow-*.whl

WORKDIR /root

# Set up CUDA variables
ENV CUDA_PATH /usr/local/cuda
ENV LD_LIBRARY_PATH /usr/local/cuda/lib64

# TensorBoard
EXPOSE 6006
# IPython
EXPOSE 8888

RUN ["/bin/bash"]
```

Then I built the image and ran it with `./docker_run_gpu.sh tf/tf`

Started tensorflow with: `python -m tensorflow.models.image.mnist.convolutional`

And I'm getting `failed call to cuInit: CUDA_ERROR_UNKNOWN` when tensorflow starts up. But this is after reporting that it `successfully opened CUDA library libcuda.so locally`.

```
I tensorflow/stream_executor/dso_loader.cc:101] successfully opened CUDA library libcublas.so.7.0 locally
I tensorflow/stream_executor/dso_loader.cc:101] successfully opened CUDA library libcudnn.so.6.5 locally
I tensorflow/stream_executor/dso_loader.cc:101] successfully opened CUDA library libcufft.so.7.0 locally
I tensorflow/stream_executor/dso_loader.cc:101] successfully opened CUDA library libcuda.so locally
I tensorflow/stream_executor/dso_loader.cc:101] successfully opened CUDA library libcurand.so.7.0 locally

I tensorflow/core/common_runtime/local_device.cc:40] Local device intra op parallelism threads: 8
modprobe: ERROR: ../libkmod/libkmod.c:556 kmod_search_moddep() could not open moddep file '/lib/modules/3.19.0-32-generic/modules.dep.bin'
E tensorflow/stream_executor/cuda/cuda_driver.cc:481] failed call to cuInit: CUDA_ERROR_UNKNOWN
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:114] retrieving CUDA diagnostic information for host: 24b008aee65f
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:121] hostname: 24b008aee65f
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:146] libcuda reported version is: Not found: was unable to find libcuda.so DSO loaded into this program
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:257] driver version file contents: """NVRM version: NVIDIA UNIX x86_64 Kernel Module  352.68  Tue Dec  1 17:24:11 PST 2015
GCC version:  gcc version 4.8.4 (Ubuntu 4.8.4-2ubuntu1~14.04)
```

**Comment by dcunited001** [comment]
i think i'm running into errors because modprobe fails in the container. 

```
modprobe: ERROR: ../libkmod/libkmod.c:556 kmod_search_moddep() could not open moddep file '/lib/modules/3.19.0-32-generic/modules.dep.bin'
```

 I also needed to add my docker user to sudo group, so I did that.  but then `/lib/modules` wasn't mapped in the container, so i fixed that with the following addition to my run script

```
export LIB_MODULES=$(\uname -r | xargs -I{} echo '-v /lib/modules/{}:/lib/modules/{}')

# ....

docker run -it $CUDA_SO $LIB_MODULES $DEVICES "$@"
```

but now, when i run `sudo modprobe nvidia` i'm getting 'modprobe: FATAL: Module nvidia not found.'

**Comment by dcunited001** [comment]
i was able to build on my local machine for now.  going to try again with docker later, but i think the issue was my host machine configuration.  for some reason the docker container is not permitted or cannot access the nvidia drivers.

**Comment by ebrevdo** [comment]
You should not need to run modprobe in the container, only on the host.
Once that runs, the /dev/nvidia* devices are accessible and you can launch
the container.

**Comment by dcunited001** [comment]
@ebrevdo k, i'm going to build my docker image once more and try again, but i believe i still get the `CUDA_ERROR_UNKNOWN` when i launch the Tensorflow MNIST example.  

From what I recall, the `/dev/nvidia*` devices were missing in the container on my last build, even though `docker_run_gpu.sh` maps those files.  furthermore, i modified the docker gpu shell script to include the following bc these files seemed to be missing or inaccessible from the container.

```
# trying to make modprobe accessible
# - making this available makes modprobe accessible
# - but modprobe still can't find nvidia
export LIB_MODULES=$(\uname -r | xargs -I{} echo '-v /lib/modules/{}:/lib/modules/{}')
```  

i will update again after the container builds.

**Comment by dcunited001** [comment]
also @ebrevdo, i've got a quick question which may save me a lot of time:

i'm working with a 30GB image set, where the images range from 50MB-150MB.  i want to use the GPU version of OpenCV, which doesn't offer python bindings.  in order to perform OpenCV modifications in TensorFlow nodes, do i need to write my own TensorFlow ops?  And if so, can I do that without rebuilding the entire project? And a TensorFlow op be written in c++ outside of the main repository and just link back into it?  Or are there build artifacts used in the main build of TensorFlow that require all ops to be available? 

Or does using OpenCV in that way kind of defeat the purpose of TensorFlow.  I plan on using `SIFT`, possibly `SURF` (since i can GPU accel), colorspace + filter + segmentation + morphology and other algorithms.  There are only a few of these algorithms directly implemented in TensorFlow.  It would seem that I could compose some of the available API functions in TensorFlow to recreate most of the algorithms I could need, but doing so would take more time.  E.G., I don't know that I could simply/efficiently replicate SIFT in TensorFlow without spending quite a few hours.  

This is my first computer vision project and it is not classification, but quantifying attributes of types of objects in images to create a neural net prediction model for the same types of images.

These are the kinds of questions that would save me lots of time if I didn't need to search for the answer.  Is there a chatroom or something where I can pose questions like this to people working with OpenCV and TensorFlow?

**Comment by dcunited001** [comment]
in addition to this, i logged my output when running `docker build . -t tf/tf`. I can upload that to a gist or something, if it would help.  i think there's something wrong with my host system.

from inside the container, `ls -ll /dev/nvidia*` gives me:

```
root@f206ea45c11e:~# ls -ll /dev/nvidia*
crw-rw-rw- 1 root root 195,   0 Dec 28 22:17 /dev/nvidia0
crw-rw-rw- 1 root root 195, 255 Dec 28 22:17 /dev/nvidiactl
```

my docker container arguments from `./docker_run_gpu.sh` include: `/lib/modules/3.19.0-32-generic:/lib/modules/3.19.0-32-generic`, though i'm not sure this is required.  it was required for me to add this to the shell script in order to make the folder accessible within the container.  now it seems to be accessible (and the nvidia & nvidia-uvm driver files are included) and I'm no longer seeing the following error when i run tensorflow:

```
modprobe: ERROR: ../libkmod/libkmod.c:556 kmod_search_moddep() could not open moddep file '/lib/modules/3.19.0-32-generic/modules.dep.bin'
```

instead, i'm seeing:

```
modprobe: FATAL: Module nvidia-uvm not found.
```

i'm still getting `CUDA_ERROR_UNKNOWN` when running MNIST Convolutional from inside the container:

```
modprobe: FATAL: Module nvidia-uvm not found.
E tensorflow/stream_executor/cuda/cuda_driver.cc:481] failed call to cuInit: CUDA_ERROR_UNKNOWN
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:114] retrieving CUDA diagnostic information for host: f206ea45c11e
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:121] hostname: f206ea45c11e
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:146] libcuda reported version is: Not found: was unable to find libcuda.so DSO loaded into this program
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:257] driver version file contents: """NVRM version: NVIDIA UNIX x86_64 Kernel Module  352.68  Tue Dec  1 17:24:11 PST 2015
GCC version:  gcc version 4.8.4 (Ubuntu 4.8.4-2ubuntu1~14.04) 
"""
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:150] kernel reported version is: 352.68
I tensorflow/core/common_runtime/gpu/gpu_init.cc:127] DMA: 
Initialized!
Epoch 0.00
Minibatch loss: 12.054, learning rate: 0.010000
Minibatch error: 90.6%
Validation error: 84.6%
```

`nvidia-smi` tells me only ~200MB of GPU RAM are being used, but my CPU's are being hammered.

**Comment by dcunited001** [comment]
furthermore, when i run `python -m tensorflow.models.image.mnist.convolutional` after i loaded the container with my use (instead of the `docker` user i created to manage the docker service) i get slightly different output, shown below.  both are in my `docker` group.  i don't see the call to or error from modprobe.   as you've said modprobe access shouldn't be required, i don't know why i'm seeing it above or why there is a difference, since both users are in the docker group.  

```
I tensorflow/stream_executor/dso_loader.cc:101] successfully opened CUDA library libcuda.so locally
I tensorflow/stream_executor/dso_loader.cc:101] successfully opened CUDA library libcurand.so.7.0 locally
Succesfully downloaded train-images-idx3-ubyte.gz 9912422 bytes.
Succesfully downloaded train-labels-idx1-ubyte.gz 28881 bytes.
Succesfully downloaded t10k-images-idx3-ubyte.gz 1648877 bytes.
Succesfully downloaded t10k-labels-idx1-ubyte.gz 4542 bytes.
Extracting data/train-images-idx3-ubyte.gz
Extracting data/train-labels-idx1-ubyte.gz
Extracting data/t10k-images-idx3-ubyte.gz
Extracting data/t10k-labels-idx1-ubyte.gz
E tensorflow/stream_executor/cuda/cuda_driver.cc:481] failed call to cuInit: CUDA_ERROR_UNKNOWN
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:114] retrieving CUDA diagnostic information for host: ebc1748c6a6f
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:121] hostname: ebc1748c6a6f
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:146] libcuda reported version is: Not found: was unable to find libcuda.so DSO loaded into this program
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:257] driver version file contents: """NVRM version: NVIDIA UNIX x86_64 Kernel Module  352.68  Tue Dec  1 17:24:11 PST 2015
GCC version:  gcc version 4.8.4 (Ubuntu 4.8.4-2ubuntu1~14.04) 
"""
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:150] kernel reported version is: 352.68
I tensorflow/core/common_runtime/gpu/gpu_init.cc:127] DMA: 
Initialized!
Epoch 0.00
Minibatch loss: 12.054, learning rate: 0.010000
Minibatch error: 90.6%
Validation error: 84.6%
```

like i said, this isn't a huge deal, but i eventually need to run my software on an amazon `g2.2xlarge` instance, which also have compute capability 3.0, with `GRID K520` GPU's.  However, i don't anticipate running into these exact problems in the cloud.  i think the problem lies in my host system configuration, so i think amazon images should be cleaner.

**Comment by jtoy** [comment]
+1

**Comment by nicksahler** [comment]
+1

**Comment by abhishekpatnia** [comment]
Hi All
I have also been trying to have a healthy docker based developer environment. I have had some success. I have a Ubuntu 14 box with TitanX GPU. Here are some docker files and a runner script which might help.

Tensorflow Dockerfile: https://github.com/abhishekpatnia/Docker/blob/master/docker_images/Dockerfile.tensorflow

Also I wanted IDE support so I created a Docker image for Pycharm which can be triggered from inside a container. For example you can use the tensorflow image as base for this dockerfile.
https://github.com/abhishekpatnia/Docker/blob/master/docker_images/Dockerfile.pycharmDE

Also have a helper script to run docker containers. This will allow you to have multiple sessions with the same container. It sets up proper ports and CUDA specific things.
https://github.com/abhishekpatnia/Docker/blob/master/bin/dockersh

**Comment by hanyunfan** [comment]
+1

**Comment by raviqqe** [comment]
+1

**Comment by girving** [comment]
@martinwicke, @aselle: Who's the best person for Docker issues?  This issue seems to have fallen through the cracks.

**Comment by PhoenixDai** [comment]
I had the same problem with running tensorflow on a Ubuntu machine after I upgraded my driver to 352.63 and 352.93. (I remember it works with 346.* but when I try to install 346.*, it installs 352.* automatically for some reason). 

I finally figured out that it's caused by permission issue. (I can run it with root) So, I changed the permission of the libcuda.so.352-63 file to executable by anyone and it works well now. 

Hope this will be helpful to those still struggling with this issue. 

I didn't try the docker one, but I guess it's also caused by permission setting.

**Comment by alextp** [comment]
@caisq is this being worked on? Or should we reassign?

**Comment by Sadrpour** [comment]
No this is not fixed ... the solution of sudo apt-get install nvidia-modprobe does not work for some folks

**Comment by caisq** [comment]
I'm not sure whether Linux on Mac falls in our scope of support. @gunan, @martinwicke , any thoughts?

**Comment by martinwicke** [comment]
It is impossible for us to test for or even reproduce, so I don't see much we can do on our end. 

That said, if someone figures it out and sends a patch with a fix, I'd be very excited to accept that. 

I've changed the labels appropriately.

**Comment by gunan** [comment]
looks like there is no nvidia-docker support on OSX
https://github.com/NVIDIA/nvidia-docker/issues/175

Therefore, this seems to be blocked for us by nvidia docker.

**Comment by amitsoni** [comment]
It was a permission issue for me. Thanks @PhoenixDai, following fixed it.

`chmod 655 /var/lib/nvidia-docker/volumes/nvidia_driver/367.57/lib64/libcuda.so.367.57`

**Comment by lsqpku** [comment]
I have chgmod 777 all libcuda.so.* files in /usr/lib, /usr/lib64 and /usr/local/cuda-8.0/, but still doesn't work. However SU can success.

**Comment by lsqpku** [comment]
issue was solved after I add s authority to /usr/bin/nvidia-modprobe.

**Comment by Gitwangshuo** [comment]
I fix it by reboot the os but now it shows agains ,Did anyone knows the root reason ?
